src/lsst/rsp/startup/services/labrunner.py

```diff
@@ -0,0 +1,698 @@
+        """Start the user lab."""
+        # The LabRunner is not actually the first thing that launches when
+        # we start a user lab.
+        #
+        # At the moment, the only Python 3 in the Lab container is the one that
+        # is part of the DM stack conda environment.
+        #
+        # So to get far enough to even start the LabRunner, we need to have
+        # already sourced the shell magic that sets up that conda env.
+        #
+        # However, even before we do that, we check the environment to see
+        # whether it needs resetting.  That's actually kind of handy to do
+        # before we start any Python process, because it's not impossible that
+        # the user could have messed up their own Python environment so badly
+        # that Python couldn't even get this far.
+        #
+        # When we split the stack Python from the Python-that-runs-JuptyerLab,
+        # there will be no need to set up the stack before launching
+        # Jupyterlab, and the likelihood that the user environment is so
+        # corrupt that the user cannot get a terminal session open to purge
+        # it themselves will lessen greatly.  At that point moving the
+        # user-environment purge into lsst-rsp may be feasible.
+        #
+
+        # Set up environment variables that we'll need either to launch the
+        # Lab or for the user's terminal environment
+        self._configure_env()
+
+        # Copy files into the user's home space.  If $HOME is not mounted
+        # and writeable, things will go wrong here.
+        self._copy_files_to_user_homedir()
+
+        # Check out notebooks, and set up git-lfs
+        self._setup_git()
+
+        # Clear EUPS cache
+        self._cmd.run("eups", "admin", "clearCache")
+
+        # Decide between interactive and noninteractive start, do
+        # things that change between those two, and launch the Lab
+        self._launch()
+
+    def _externalize(self, setting: str) -> str:
+        # We build multiple settings by concatenating `EXTERNAL_INSTANCE_URL`
+        # with some other string.  Make this robust by accepting either
+        # a slash or no slash on the end of `EXTERNAL_INSTANCE_URL` and
+        # returning a string with exactly one slash as a separator between
+        # the external URL and the setting.
+        ext_url = self._env["EXTERNAL_INSTANCE_URL"]
+        return f"{ext_url.strip('/')}/{setting.lstrip('/')}"
+
+    #
+    # Next up, a big block of setting up our subprocess environment.
+    #
+    def _configure_env(self) -> None:
+        self._logger.debug("Configuring environment for JupyterLab process")
+
+        self._set_user()
+
+        self._set_cpu_variables()
+
+        self._set_image_digest()
+
+        self._expand_panda_tilde()
+
+        self._set_launch_params()
+
+        self._set_firefly_variables()
+
+        self._force_jupyter_prefer_env_path_false()
+
+        self._set_butler_credential_variables()
+
```

Comment:
Oh, sorry, without the comments it's totally fine to drop the blank lines as well. My note about the blank lines was just if you wanted to keep the comments, since it's the code -> comment -> code pattern that I was finding visually hard to read (specifically to distinguish the comments from the code).